### PR TITLE
doc: improve error output in notation key and notation cert

### PR DESCRIPTION
**Pull request by FeynmanZhou**
Update CLI spec to improve error output in `notation key` and `notation cert` and resolve the UX problem in #604

### CODE REVIEW

File: specs/commandline/key.md
Comment (shizhMSFT) [inline specs/commandline/key.md:51]: Do we want to have an alias such as
```
Usage:
  notation key remove [flags] <key_name>...

Aliases:
  remove, delete
```

File: specs/commandline/key.md
Comment (shizhMSFT) [inline specs/commandline/key.md:127]: What's the output if the key is associated with a plugin?

File: specs/commandline/certificate.md
Comment (patrickzheng200) [inline specs/commandline/certificate.md:220]: This Error message is triggered/created automatically by Cobra:
`Error: required flag(s) "store", "type" not set`

File: specs/commandline/key.md
Comment (patrickzheng200) [inline specs/commandline/key.md:127]: 1. Since we do not support local key, the key created by `generate-test` becomes the only use case of this error message.
2. `You still need to delete the certificate from <cert_path>.` We can show this message only when the key is created by `generate-test`. Because in production, `decide not to Sign with a key anymore` is different from `decide not to trust a certificate in Verification anymore`.

File: specs/commandline/certificate.md
Comment (FeynmanZhou) [reply specs/commandline/certificate.md:220]: Is the error output msg customizable?

File: specs/commandline/key.md
Comment (FeynmanZhou) [reply specs/commandline/key.md:127]: Agree 1 and 2. This is against locally creating the test key/cert by `generate-test`.
I will add a prerequisite to explain this scenario in Spec. We also need to differentiate scenarios and explicitly tell users that the remote key/cert can not be deleted by Notation.

File: specs/commandline/key.md
Comment (FeynmanZhou) [reply specs/commandline/key.md:51]: I would not suggest adding an alias here since `Delete` has a different meaning from `Remove`. `Remove` is more accurate in the context of `notation key`. 

I am open to this question. @toddysm @iamsamirzon @yizha1 @vaninrao10

Comment (yizha1) [reply specs/commandline/key.md:51]: My opinion is No. So there are no aliases for `notation key remove` command

File: specs/commandline/certificate.md
Comment (patrickzheng200) [reply specs/commandline/certificate.md:220]: > Is the error output msg customizable?

As far as I know, no, it cannot be customized.

File: specs/commandline/key.md
Comment (yizha1) [inline specs/commandline/key.md:118]: Is `specific` better?

File: specs/commandline/key.md
Comment (patrickzheng200) [reply specs/commandline/key.md:127]: Yes, and as we discussed, we should introduce `notation generate-test` command to decouple `generate-test` with `cert`. This can be done between RC.4 and v1.0.

Comment (yizha1) [reply specs/commandline/key.md:127]: I agree with Patrick

Comment (yizha1) [reply specs/commandline/key.md:127]: There is a case that if the key to be removed is marked as default key. The output message should be improved as well.

**Review by yizha1 — COMMENTED** [review]
Some comments left.

File: specs/commandline/key.md
Comment (FeynmanZhou) [reply specs/commandline/key.md:127]: @patrickzheng200 I will create another issue to track this new feature request.

Comment (FeynmanZhou) [reply specs/commandline/key.md:127]: @yizha1 @patrickzheng200  If a default key is removed from the signing key list, naturally, it means there is no default key anymore. So it's not necessary to tell users the key `xxx` is unmarked as default in the output msg.

File: specs/commandline/key.md
Comment (FeynmanZhou) [reply specs/commandline/key.md:118]: No. I think `specified` points to a particular thing here. See https://www.askdifference.com/specific-vs-specified/

File: specs/commandline/key.md
Comment (patrickzheng200) [reply specs/commandline/key.md:127]: @FeynmanZhou On a second thought, this actually is not a simple change. When deleting keys, we are deleting them by key name. From key name itself, we cannot tell if that's the key that been generated by `notation cert generate-test` command. In order to do it, we need to also make changes in notation-go library. Since we have agreed to introduce `notation generate-test` command in next release, I suggest we just make a minor change on the error message here in RC.4, something like:
`Removed test key <key_name> from Notation signing key list. The source key file is not deleted.`

Comment (yizha1) [reply specs/commandline/key.md:127]: @patrickzheng200 When using `notation key list`, `key path` is also printed out, maybe we can print out the `The source <key path> is not deleted. `  If it also takes more time, I agree with your proposal. There is also `cert path` in the output message of `notation key list` and seems only useful for local testing keys. We can improve this later with the new `notation generate-test` command.

Comment (patrickzheng200) [reply specs/commandline/key.md:127]: > @patrickzheng200 When using `notation key list`, `key path` is also printed out, maybe we can print out the `The source <key path> is not deleted. ` If it also takes more time, I agree with your proposal. There is also `cert path` in the output message of `notation key list` and seems only useful for local testing keys. We can improve this later with the new `notation generate-test` command.

@yizha1 Good idea, but I think it won't work on Deletion. Here's the reason: `notation key list` reads in the `signingkeys.json` file and print out the keys one by one. When deleting one or more keys, we delete by the key names (the core deletion logic is in notation-go). At the point of printing out, those keys have already been deleted from `signingkeys.json`. In other words, we cannot find the deleted key names anymore by just going through the file again. A workaround would be saving a copy of the whole key list before doing any Deletion. When printing out, we go through that pre-load list. But IMO, this isn't something that we should do just for some print outs. Therefore, I still suggest we use the message: `Removed test key <key_name> from Notation signing key list. The source key file is not deleted.` and introduce `notation generate-test` command in the next release to solve the issue completely.

Comment (FeynmanZhou) [reply specs/commandline/key.md:127]: @patrickzheng200 Agree. We could improve it iteratively and aim to fix it by introducing a new command in v1.0.0 completely.

File: specs/commandline/key.md
Comment (toddysm) [reply specs/commandline/key.md:51]: I believe, we had this discussion with either @iamsamirzon or @priteshbandi in the past. We need to be clear what is the meaning of `delete` and `remove`. It seems different people have different expectation. As a non-English speaker my expectation is similar to @shizhMSFT, i.e. I expect the key (and associated assets) to be gone completely. Though, the implementation right now is that the key is only "deleted" from the list of keys but not removed from the file system and associated certs are also not removed. As an experience, I would expect the command to be intelligent enough to give choices to do all of the above and not have separate command to "remove".

*Note:* We are talking about `key delete` here as far as I can see from @shizhMSFT comment.

File: specs/commandline/key.md
Comment (toddysm) [reply specs/commandline/key.md:127]: Are we renaming the command from `delete` to `remove`? This will be a breaking change - isn't it?

File: specs/commandline/key.md
Comment (shizhMSFT) [reply specs/commandline/key.md:51]: `remove` v.s.` delete`:
- `remove` is about removing something from a place and **can** be easily recovered. For instance, remove a song from a playlist.
- `delete` is about removing something from a place and **cannot** be easily recovered. For instance, delete a song from a music library.

In `notation key`, all keys are referenced. Therefore, `key remove` is more suitable. Removing the key from list will not delete the key from the storage.
In `notation cert`, all certs are stored as files. Therefore, `cert delete` is more suitable. Deleting the cert will delete the cert file from the disk and cannot be easily recovered.

However, as @toddysm mentioned, non-English speakers mix the term `remove` and `delete`. That's why I initially proposed to add `delete` as an alias to `remove`. Of course, we need to clearly state the key is **removed** in the response.

On the other hand, completely removing the key using `notation` is technically not possible unless a `delete key` protocol is introduced for plugins.

File: specs/commandline/key.md
Comment (byronchien) [inline specs/commandline/key.md:124]: grammar nit:
```suggestion
- Upon successful execution, the output message is printed out as below when removing a key created by notation for local testing purposes:
```

or can be reworded

```suggestion
- Upon successful removal of a local testing key created by notation, the output message is printed out as follows:
```

File: specs/commandline/key.md
Comment (byronchien) [inline specs/commandline/key.md:129]: grammar nit:
```suggestion
- Upon successful execution, the output message is printed out as below when removing a key associated with a KMS:
```

or can be reworded

```suggestion
- Upon successful removal of a key associated with a KMS, the output message is printed out as follows:
```

File: specs/commandline/key.md
Comment (byronchien) [inline specs/commandline/key.md:141]: do we need an example output for removing multiple keys? even if it's just
```shell
key_name_1
key_name_2
```

File: specs/commandline/key.md
Comment (patrickzheng200) [inline specs/commandline/key.md:127]:
```suggestion
Removed <key_name> from Notation signing key list. The source key file is not deleted.
```

File: specs/commandline/key.md
Comment (yizha1) [reply specs/commandline/key.md:51]: My two cents:

It seems a practice that `add` and `remove` are a pair; `create` and `delete` are a pair. This is also another reason `notation key` and `notation cert` commands cause confusion.

Since we don't create any resources (key and certificate) except the `notation cert generate-test`. Using `add` and `remove` are more appropriate in most scenarios for `notation key` and `notation cert` commands. 

We could consider a different command for local key or certificate later.

Comment (yizha1) [reply specs/commandline/key.md:51]: We can fix the error and output messages first to make current usage clear enough to our users.

Comment (FeynmanZhou) [reply specs/commandline/key.md:51]: Per the discuss in the community meeting on Apr 3, we agreed that change `notation key delete` to `notation key remove` and use `delete` as an alias of `remove`.

File: specs/commandline/key.md
Comment (FeynmanZhou) [reply specs/commandline/key.md]: Accepted.

File: specs/commandline/key.md
Comment (FeynmanZhou) [reply specs/commandline/key.md]: Accepted and added a new output example for removing multiple keys.

File: specs/commandline/key.md
Comment (FeynmanZhou) [reply specs/commandline/key.md]: Updated

File: specs/commandline/key.md
Comment (patrickzheng200) [reply specs/commandline/key.md]: @FeynmanZhou I thought we agreed on this 2 weeks ago: https://github.com/notaryproject/notation/pull/606#discussion_r1154298416.

Comment (FeynmanZhou) [reply specs/commandline/key.md]: @patrickzheng200 It's fine that we do not print the `<key_path>` in the output for this release. Currently, there is a [guide](https://github.com/notaryproject/notaryproject.dev/pull/164/files#diff-ebf28445e02ae412163e2fde43e76bee51e880abab8449ac8b4b863fec8d4354R29 ) to tell users to clean up those source key and cert files. We can determine if we should design a feature like `notation generate-test` or support local signing key scenario.

File: specs/commandline/key.md
Comment (patrickzheng200) [inline specs/commandline/key.md:132]:
```suggestion
Removed <key_name> from Notation signing key list. The source key files still exist.
```
We should mention the source key files still exist for KMS scenario as well. Otherwise, users might misunderstand that the key is also removed from their KMS.

File: specs/commandline/key.md
Comment (patrickzheng200) [inline specs/commandline/key.md:144]:
```suggestion
Removed the following keys from Notation signing key list. The source key files still exist.
```
When deleting a key, telling it's a local key or a key stored in a KMS is as tricky as printing out the local key path. Actually, if we could tell a key is a local key, then we are able to print its path out. Both cases need changes on notation CLI + notation-go library.

File: specs/commandline/key.md
Comment (patrickzheng200) [inline specs/commandline/key.md:127]:
```suggestion
Removed <key_name> from Notation signing key list. The source key file still exists.
```
same here.

File: specs/commandline/key.md
Comment (FeynmanZhou) [reply specs/commandline/key.md:144]: Got it. Keeping minimum changes on the CLI outputs is acceptable without adding additional logic in notation-go lib this time. We could improve this experience thoroughly after v1.0.0 by proposing a feature to clean up local testing keys.

**Review by patrickzheng200 — APPROVED** [review]
LGTM

**Review by patrickzheng200 — DISMISSED** [review]
LGTM

File: specs/commandline/key.md
Comment (yizha1) [inline specs/commandline/key.md:127]: If the source key is a remote key, it may not be a key file? How about "The source key still exists."

File: specs/commandline/key.md
Comment (yizha1) [inline specs/commandline/key.md:132]: If the source key is a remote key, it may not be a file. How about "The source key still exists."  Maybe we can use this for local key as well.

File: specs/commandline/key.md
Comment (yizha1) [inline specs/commandline/key.md:127]: Do we need to say something if the removed key is a default key?

Comment (FeynmanZhou) [reply specs/commandline/key.md:127]: @yizha1  I think it's not necessary to mention that when deleting a key whether it is a default key or not. https://github.com/notaryproject/notation/pull/606#discussion_r1154195936

**Review by patrickzheng200 — APPROVED** [review]
LGTM

**Review by yizha1 — APPROVED** [review]
LGTM but IANAM

File: specs/commandline/certificate.md
Comment (JeyJeyGao) [inline specs/commandline/certificate.md:211]:
```suggestion
A prompt is displayed, asking the user to confirm the deletion. Upon successful deletion, the specific certificate is deleted from the trust store named `<name>` of type `<type>`. The output message is printed out as following:
```

**Review by priteshbandi — APPROVED** [review]
LGTM

**Comment by FeynmanZhou** [comment]
Hi @notaryproject/notaryproject-notation-maintainers , I resolved @JeyJeyGao 's comment. Could you pls review it again? Thanks

**Review by patrickzheng200 — APPROVED** [review]
LGTM

**Comment by FeynmanZhou** [comment]
@priteshbandi There was a small update after your last approval. Could you re-approve this PR? Thanks!

**Review by priteshbandi — APPROVED** [review]
LGTM

**Pull request merged**
